File: gpio4j/src/main/java/org/overworld/gpio4j/serial/wire/Message.java
package org.overworld.gpio4j.serial.wire;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.IOException;

import org.overworld.gpio4j.serial.MessageType;

public interface Message {

	public Message parse(ByteArrayInputStream wireData) throws IOException;
	
	public ByteArrayOutputStream wireData();
	
	public MessageType getType();
}

File: gpioBoardDriver/serial/Message.cpp
#include "Message.h"

class Message {
public:
    virtual ~Message() = default;

    virtual std::unique_ptr<Message> parse(std::istream& wireData) = 0;
    
    virtual std::vector<uint8_t> wireData() const = 0;
    
    virtual MessageType getType() const = 0;
};
